core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -583,4 +413,216 @@ protected boolean runSerially() {
+                final long diskPair = onDiskWalIDSegmentIDPairs.get(i);
+                final int walId = Numbers.decodeHighInt(diskPair);
+                final int segmentKey = Numbers.decodeLowInt(diskPair);
+                final int segmentId = decodeSegmentId(segmentKey);
+                final boolean hasPendingTasks = decodePendingTasks(segmentKey);
+                final boolean segmentLocked = decodeLocked(segmentKey);
+
+                if (hasPendingTasks) {
+                    continue;
+                }
+
+                final int nextToApplySegmentId = nextToApply.get(walId);  // -1 if not found
+
+                if (nextToApplySegmentId != -1) {
+                    // WAL is in use, there are committed transactions in Sequencer
+                    if (nextToApplySegmentId > segmentId) {
+                        // Segment already applied.
+                        logDebugInfo();
+                        deleter.deleteSegmentDirectory(walId, segmentId);
+                    }
+                } else if (!segmentLocked) {
+                    // If the sequencer doesn't track any unapplied changes for this walId and the walId is unlocked,
+                    // then we can delete the whole WAL directory, unless any of its segments track pending tasks.
+                    if (!lockedWalIDSet.contains(walId) && !pendingWalIDSet.contains(walId)) {
+                        logDebugInfo();
+                        deleter.deleteWalDirectory(walId);
+                    } else {
+                        // WAL is locked but the segment in it is not in outstanding commits, delete it.
+                        // Segment with ID of WalUtils.SEG_NONE_ID is the WAL directory itself: only delete it if it's unlocked.
+                        if (segmentId != WalUtils.SEG_NONE_ID) {
+                            logDebugInfo();
+                            deleter.deleteSegmentDirectory(walId, segmentId);
+                        }
+                    }
+                }
+            }
+        }
+
+        public void trackDiscoveredSegment(int walId, int segmentId, boolean pendingTasks, boolean locked) {
+            if (pendingTasks) {
+                pendingWalIDSet.add(walId);
+            }
+            final int segmentKey = encodeSegmentKey(segmentId, pendingTasks, locked);
+            onDiskWalIDSegmentIDPairs.add(Numbers.encodeLowHighInts(segmentKey, walId));
+        }
+
+        public void trackDiscoveredWal(int walId, boolean walInUse) {
+            if (walInUse) {
+                lockedWalIDSet.add(walId);
+            }
+            // This is to track WAL directory itself.
+            onDiskWalIDSegmentIDPairs.add(Numbers.encodeLowHighInts(encodeNoSegment(), walId));
+        }
+
+        public void trackNextToApplySegment(int walId, int segmentId) {
+            final int currentNextToApply = nextToApply.get(walId);  // -1 if not found
```

Comment:
there is a faster method to put-if-not-exists, see `keyIndex()`

Reply:
Thanks for the tip, will change.

Reply:
Done.